iModelCore/ECPresentation/Source/Content/PropertyInfoStore.cpp

```diff
@@ -476,25 +488,72 @@ Utf8String PropertyInfoStore::GetLabelOverride(ECPropertyCR prop, ECClassCR ecCl
     return "";
     }
 
+/*---------------------------------------------------------------------------------**//**
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+CategoryOverrideInfo const* PropertyInfoStore::GetCategoryOverride(PropertyCategoryIdentifier const& id, PropertyCategorySpecificationsList const& categorySpecs) const
+    {
+    auto override = m_categoryOverridesCache.find(id);
+    if (override != m_categoryOverridesCache.end())
+        return &override->second;
+
+    // create new category override and insert to cache
+    bvector<std::shared_ptr<PropertyCategoryRef>> categories;
+    if (SUCCESS == CreatePropertyCategoryFromSpec(categories, id, categorySpecs))
+        {
+        auto insertedElement = m_categoryOverridesCache.insert({ id, categories });
+        return &insertedElement.first->second;
+        }
+    return nullptr;
+    }
+
 /*-----------------------------------------------------------------------------**//**
 * @bsimethod
 +---------------+---------------+---------------+---------------+-----------+------*/
-std::unique_ptr<CategoryOverrideInfo const> PropertyInfoStore::GetCategoryOverride(ECPropertyCR prop, ECClassCR ecClass, PropertySpecificationCP customOverride, PropertyCategorySpecificationsList const* scopeCategorySpecs) const
+CategoryOverrideInfo const* PropertyInfoStore::GetCategoryOverride(ECPropertyCR prop, ECClassCR ecClass, PropertySpecificationCP customOverride, PropertyCategorySpecificationsList const* scopeCategorySpecs) const
     {
-    auto const& ovr = GetOverrides(ecClass).GetCategoryOverride(prop);
+    auto const& classOverrides = GetOverrides(ecClass);
+    auto const& categoryOverride = classOverrides.GetCategoryOverride(prop);
     bool hasPropertySpecCategoryOverride = customOverride && scopeCategorySpecs && nullptr != customOverride->GetCategoryId();
 
     if (hasPropertySpecCategoryOverride)
         {
-        if (ovr.IsValid() && ovr.Value().priority > customOverride->GetOverridesPriority())
-            return std::make_unique<CategoryOverrideInfo>(*ovr.Value().value);
+        if (categoryOverride.IsValid() && categoryOverride.Value().priority > customOverride->GetOverridesPriority())
+            return GetCategoryOverride(*categoryOverride.Value().value, classOverrides.GetAvailableCategories());
 
-        return CreateCategoryOverride(*customOverride->GetCategoryId(), *scopeCategorySpecs);
+        // try to get categories from local scope
+        auto overrideFromLocalScope = GetCategoryOverride(*customOverride->GetCategoryId(), *scopeCategorySpecs);

```

Comment:
just one question - why do we need `scopeCategorySpecs` here but not in `GetCategoryOverride(ECClassCP, CalculatedPropertiesSpecificationCR)`?

Reply:
Scope category specs are needed when we create content with related properties and we have a category override for related property when this category is specified locally. For example, we have a content modifier:
![image](https://github.com/iTwin/imodel-native/assets/75395229/fcdbce54-85f3-4230-80c3-9b083abf0353)

In this example `custom-category` will be assigned as a category for class A. 
So when we will come to this function `GetCategoryOverride(ECPropertyCR, ECClassCR ecClass, PropertySpecificationCP, PropertyCategorySpecificationsList const* scopeCategorySpecs)` and try to get category override for related property, ecClass will be B and we will try to get categories list for a class B with a function `GetOverrides(ecClass)` and there will be no categories. So we will get them from this `scopeCategorySpecs` list that contains categories from local scope..

With `calculated properties` we are not going to face this situation, because calculated properties are always specified in content modifiers or content specifications, so we will always get categories with  the `GetOverrides(ecClass)`.

I hope I explained it.